torch/_dynamo/output_graph.py

```diff
@@ -2229,6 +2198,274 @@ def maybe_lift_tracked_freevar_to_input(self, arg):
+        elif t.layout in {torch.sparse_csc, torch.sparse_bsc}:
+            self._lift_symbols_in_tensor(t.ccol_indices(), src)
+            self._lift_symbols_in_tensor(t.row_indices(), src)
+        if is_traceable_wrapper_subclass(t):
+            attrs, ctx = t.__tensor_flatten__()
+            for attr in attrs:
+                inner_t = getattr(t, attr)
+                self._lift_symbols_in_tensor(
+                    inner_t, AttrSource(src, attr) if src is not None else None
+                )
+
+    # Lookup the proxy in current tracer for each symbol in expressions of s,
+    # See Note [Auto lift basic free symbols when create_graph_input]
+    def lookup_bounded_symbols(
+        self, s: torch.SymInt
+    ) -> Tuple[Dict[sympy.Symbol, torch.fx.Proxy], Set[sympy.Symbol]]:
+        free_symbols = s.node.expr.free_symbols
+        if len(free_symbols) == 0:
+            return {}, set()
+
+        bounded = {}
+        to_be_bound = set()
+        for s0 in free_symbols:
+            if s0 not in self.bound_symbols:
+                to_be_bound.add(s0)
+                continue
+
+            proxy = self.bound_symbols[s0]
+            proxy = proxy() if isinstance(proxy, SubgraphTracer.LazyProxy) else proxy
+            assert (
+                proxy.tracer is self
+            ), f"The proxy of symbol {s0} doesn't belong to current tracer."
+            self.bound_symbols[s0] = proxy
+            bounded[s0] = proxy
+        return bounded, to_be_bound
+
+    # See NOTE: [Auto lift basic free symbols when create_graph_input]
+    def _lift_symbols_in_symint(
+        self, s: Union[int, torch.SymInt], source: Optional[Source]
+    ):
+        if not is_symbolic(s):
+            return
+
+        assert isinstance(s, torch.SymInt)
+        self_bounded, self_to_be_bound = self.lookup_bounded_symbols(s)
+        if len(self_to_be_bound) == 0:
+            return self_bounded
+
+        # For subgraph
+        if self.parent is not None:
+            # Recursively lift symbols in symint until top-level.
+            parent_bounded = self.parent._lift_symbols_in_symint(s, source)
+            for s0 in self_to_be_bound:
+                parent_proxy = parent_bounded[s0]
+                example_val = parent_proxy.node.meta["example_value"]
+                assert isinstance(example_val, torch.SymInt)
+                ph = self.create_graph_input(
+                    str(s0),
+                    type(example_val),
+                    example_val,
+                    before=True,
+                    source=source,
+                )
+                log.debug(
+                    "_lift_symbols_in_symint %s from %s at debug_level %s",
+                    s0,
+                    source.name() if source is not None else "subgraph inputs",
+                    self.debug_level,
+                )
+                self.bound_symbols[s0] = ph
+                self.lifted_freevars[parent_proxy] = ph
```

Comment:
It's a free symbol, not a freevar I think, so this is a little weird to me. Is there a mechanical reason to update this here?

Reply:
Is it for duplication of SymInts? If so, I thought we duplicate already using bound_symbols?